core/templates/dev/head/filters.js

```diff
@@ -345,26 +345,56 @@ oppia.filter('normalizeWhitespacePunctuationAndCase', [function() {
   };
 }]);
 
-// Note that this filter removes additional new lines or <p><br></p> tags
-// at the end of the string.
-oppia.filter('removeExtraLines', [function() {
+// Filter that targets to remove occurrence of blank lines, empty paragraphs
+// and spaces at the end of string. It'll remove <p></p>, <p><br></p>, spaces
+// and &nbsp; occurrences from the end of the string.
+oppia.filter('removeExtraLinesAndSpace', [function() {
   return function(string) {
     if (!angular.isString(string)) {
       return string;
     }
     var BLANK_LINES_TEXT = '<p><br></p>';
     var EMPTY_PARA_TEXT = '<p></p>';
+    var BLANK_CHAR = ' ';
+    var BREAK_LINE_TAG = '<br>';
+    var SPACE_ENTITY = '&nbsp;';
     while (1) {
-      var endIndex = string.length;
-      var bStr = string.substring(endIndex - BLANK_LINES_TEXT.length, endIndex);
-      var pStr = string.substring(endIndex - EMPTY_PARA_TEXT.length, endIndex);
-      if (bStr === BLANK_LINES_TEXT) {
-        string = string.substring(0, endIndex - BLANK_LINES_TEXT.length);
-      } else if (pStr === EMPTY_PARA_TEXT) {
-        string = string.substring(0, endIndex - EMPTY_PARA_TEXT.length);
-      } else {
+      var lastIndexOfClosingTag = string.lastIndexOf('<');
```

Comment:
I'm worried about this. It's making some assumptions, e.g. that the final tag is `</p>` and not something like `<br>`.

In general I also think that this function is way too complicated. Putting lots of stuff in `while(1)` loops is not a great pattern. If you make a mistake (and that's easy to do in a function as complex as this), you're gonna crash the user's browser tab.

I would like to suggest a different approach. Could you create a jQuery element with the content, and then remove _nodes_ from the end of that element without using while loops? It would be easier to comprehend, and, importantly, it does not use raw string manipulation. String manipulation of HTML is not a good pattern to have as it introduces security risks. See e.g. line 132 and subsequent lines of services/RteHelperService.js (in convertHtmlToRte) for a cleaner way to do this.

I acknowledge that the previous implementation of this used string manipulation and that you were probably riffing off of that; the old implementation was not a great pattern as well. I'm going to revert this PR for safety, and I recommend creating a new one to address the original issue.